src/maxtree.jl

```
@@ -0,0 +1,904 @@
+                            all_below_min = zero(eltype(output)))
+    # should have been already checked by higher-level functions
+    @assert size(output) == size(image)
+    @assert size(image) == size(maxtree)
+    @assert length(attrs) == length(maxtree)
+
+    p_root = root_index(maxtree)
+    # if p_root is below min_attr, then all others are as well
+    (attrs[p_root] < min_attr) && return fill!(output, all_below_min)
+    output[p_root] = image[p_root]
+    parentindices = maxtree.parentindices
+
+    for p in maxtree.traverse
+        q = parentindices[p]
+        # this means p is not canonical
+        # in other words, it has a parent that has the
+        # same image value.
+        if image[p] == image[q]
+            output[p] = output[q]
+            continue
+        end
+        # if attribute is below the threshold, stop at the lower parent,
+        # otherwise continue to reconstruct the image
+        output[p] = attrs[p] < min_attr ? output[q] : image[p]
+    end
+
+    return output
+end
+
+function check_output_image(output::GenericGrayImage, image::GenericGrayImage)
+    (size(output) == size(image)) ||
+        throw(DimensionMismatch("The sizes of the output and the input image do not match"))
+end
+
+# checks if the provided maxtree is compatible with the given options
+# or builds the new maxtree if none was given
+function check_maxtree(maxtree::Union{MaxTree, Nothing},
+                       image::GenericGrayImage;
+                       connectivity::Integer=1,
+                       rev::Bool=false)
+    (maxtree === nothing) && return MaxTree(image, connectivity=connectivity, rev=rev)
+    (size(maxtree) == size(image)) ||
+        throw(DimensionMismatch("The sizes of the max-tree and the input image do not match"))
+    (maxtree.rev == rev) ||
+        throw(ArgumentError("The traversal order of the given max-tree is different from the requested one"))
+    return maxtree
```

There's a bug here:

```julia
mtree = MaxTree(img, connectivity=2, rev=false)
diameter_opening(img, connectivity=1, maxtree=mtree)
```

---

It's explained in the docstring that when `maxtree` is specified, the `connectivity` option is ignored.
To fix that, we can store the connectivity in the `MaxTree`, and use `connectivity=nothing` as the default methods' option (and substitute it with 1 internally), but I find it a bit cumbersome for such minor problem + I don't like hiding the real default values from the function signature.

---

This should not be ignored silently and should raise an error an error an error!

---

> This should not be ignored silently and should raise an error an error an error!

As you have noted, it happens only when the call looks like
```julia
diameter_opening(img, connectivity=1, maxtree=mtree)
```
which doesn't make much sense to me: in 99% of the cases the users would not create *MaxTree* explicitly (since in 99% of the cases it's only required for a single operation), and when they, for some reason, do, they would not specify `connectivity` in addition. And when they do specify, there's an explicit note in the docstring that *maxtree* parameter makes precedence.

---

Then how about

```julia
mtree = MaxTree(img, rev=false)
diameter_opening(img, connectivity=2, maxtree=mtree)
```

and I bet 99% users (including me) don't really read the docstrings that are longer than 20 lines :P

---

> Then how about `diameter_opening(img, connectivity=2, maxtree=mtree)`

It's the same: I'm not sure the person who makes such call understands what she/he wants to achieve, but `maxtree=` takes precedence.

> and I bet 99% users (including me) don't really read the docstrings that are longer than 20 lines :P

Maybe, but it's 99% from 1% (those who tackle with `maxtree=` param) :)

---

I'm not very satisfied with your explanations and I strongly suggest this be fixed and tested. Currently, there're no tests for this.